File: gateway-ui/src/machineIcons.ts
export type MachineIconId =
  | 'os-macos'
  | 'os-linux'
  | 'os-windows'
  | 'device-mini'
  | 'device-laptop'
  | 'device-desktop'
  | 'device-server'
  | 'device-gpu'
  | 'device-cloud';

export type MachineIconOption = {
  id: MachineIconId;
  label: string;
  icon: string;
  group: 'OS' | 'Machine';
};

export const machineIconOptions: MachineIconOption[] = [
  { id: 'os-macos', label: 'macOS', icon: 'fa-brands fa-apple', group: 'OS' },
  { id: 'os-linux', label: 'Linux', icon: 'fa-brands fa-linux', group: 'OS' },
  { id: 'os-windows', label: 'Windows', icon: 'fa-brands fa-windows', group: 'OS' },
  { id: 'device-mini', label: 'Mini', icon: 'fa-solid fa-cube', group: 'Machine' },
  { id: 'device-laptop', label: 'Laptop', icon: 'fa-solid fa-laptop', group: 'Machine' },
  { id: 'device-desktop', label: 'Desktop', icon: 'fa-solid fa-desktop', group: 'Machine' },
  { id: 'device-server', label: 'Server', icon: 'fa-solid fa-server', group: 'Machine' },
  { id: 'device-gpu', label: 'GPU', icon: 'fa-solid fa-microchip', group: 'Machine' },
  { id: 'device-cloud', label: 'Cloud', icon: 'fa-solid fa-cloud', group: 'Machine' },
];

const validMachineIconIds = new Set(machineIconOptions.map((option) => option.id));

export function sanitizeMachineIconIds(value: unknown): MachineIconId[] {
  if (!Array.isArray(value)) return [];
  const seen = new Set<string>();
  const icons: MachineIconId[] = [];
  for (const item of value) {
    if (typeof item !== 'string' || !validMachineIconIds.has(item as MachineIconId) || seen.has(item)) continue;
    seen.add(item);
    icons.push(item as MachineIconId);
  }
  return icons;
}

export function machineIconClass(id: MachineIconId) {
  return machineIconOptions.find((option) => option.id === id)?.icon ?? 'fa-solid fa-circle';
}

File: So Much For Subtlety/Sources/SoMuchForSubtlety/MachineIcons.swift
import SwiftUI

/// Selectable machine icons, mirroring `gateway-ui/src/machineIcons.ts` (FontAwesome
/// there → SF Symbols here).
struct MachineIconOption: Identifiable {
    enum Group: String, CaseIterable { case os = "OS", machine = "Machine" }
    let id: String
    let label: String
    let symbol: String
    let group: Group
}

enum MachineIcons {
    static let options: [MachineIconOption] = [
        .init(id: "os-macos", label: "macOS", symbol: "apple.logo", group: .os),
        .init(id: "os-linux", label: "Linux", symbol: "terminal", group: .os),
        .init(id: "os-windows", label: "Windows", symbol: "macwindow", group: .os),
        .init(id: "device-mini", label: "Mini", symbol: "cube", group: .machine),
        .init(id: "device-laptop", label: "Laptop", symbol: "laptopcomputer", group: .machine),
        .init(id: "device-desktop", label: "Desktop", symbol: "desktopcomputer", group: .machine),
        .init(id: "device-server", label: "Server", symbol: "server.rack", group: .machine),
        .init(id: "device-gpu", label: "GPU", symbol: "memorychip", group: .machine),
        .init(id: "device-cloud", label: "Cloud", symbol: "cloud", group: .machine),
    ]

    static func symbol(for id: String) -> String {
        options.first { $0.id == id }?.symbol ?? "circle"
    }
}

struct MachineIconGlyph: View {
    let id: String
    var selected = false
    var compact = false

    var body: some View {
        Image(systemName: MachineIcons.symbol(for: id))
            .font(.system(size: compact ? 10 : 12, weight: .medium))
            .symbolRenderingMode(.monochrome)
            .frame(width: compact ? 20 : 26, height: compact ? 18 : 22)
            .background(selected ? Color.accentColor.opacity(0.28) : Color.secondary.opacity(0.12),
                        in: RoundedRectangle(cornerRadius: compact ? 4 : 5))
    }
}
